ref: Switch to magic string for logger statements

lforst opened the pull request:
Switches logger call guards from important constant to magic string.

#### How?

- Remove `flags.ts` files from all packages (except angular - we can't use rollup there)
- Add global definition for `__DEBUG_BUILD__` in `@sentry/types`
- Replace all instances of `IS_DEBUG_BUILD` with `__DEBUG_BUILD__`
- Change rollup configuration to replace magic string with boolean for browser bundles.
- Change rollup configuration to replace magic string with safe statement `(typeof __SENTRY_DEBUG__ === 'undefined' || __SENTRY_DEBUG__)` for cjs/esm builds. The magic string in this statement can in return be replaced by downstream users to enable tree shaking of Sentry SDK logger code.

#### Why?

The resulting code we produce is much more simple for bundlers to tree shake, because they don't have to insert the const across file boundaries.

Ref: https://getsentry.atlassian.net/browse/WEB-847

lforst commented on packages/utils/src/logger.ts:
```diff
@@ -1,11 +1,7 @@
 import { getGlobalObject, getGlobalSingleton } from './global';
 
-// TODO: Implement different loggers for different environments
-const global = getGlobalObject<Window | NodeJS.Global>();
```
This caused node 8 tests to fail. Defining a variable "global" is probably not a good idea because of namespace collisions so I renamed it in some instances and moved it into local scope here.

lobsterkatie replied:
Given that's how we do it across the the SDK, I'd rather keep that consistent and just fix the tests. This, in theory, is the problem the const-to-var transformer is supposed to solve. Lemme take a look at the tests and see if I can figure out why that's not working.

lforst replied:
If it helps me getting this PR merged I'm happy to make this change in all of the SDK.

I genuinely believe declaring a variable called `global` is a bad idea. It would be the same as doing `const console = ...` or `const setTimeout = ...`. Feels weird doesn't it? I never like the argument "because we always did it that way so we must continue with it".

Also, I don't think the tests are at fault here. They're probably doing the right thing ,because they are now surfacing this. A transformer should IMO not be necessary for stuff like this. It's solving a problem that shouldn't even exist.

lobsterkatie replied:
> I genuinely believe declaring a variable called global is a bad idea. It would be the same as doing const console = ... or const setTimeout = .... Feels weird doesn't it? I never like the argument "because we always did it that way so we must continue with it".

As a general matter, I agree with you. In this case, though, it's not just "it's what we've always done." The reason we do it that way is that most of the time, the thing we're talking about _is_ `global`, and so calling it `global` is the simplest path for the reader. But we have to do the `getGlobalObject` dance, IIRC because we can't (or, at least, didn't used to be able to) use `globalThis`. Lemme see if that still applies...

UPDATE: It seems we took `globalThis` out only because we were [trying to support versions of TS and Angular we no longer support](https://github.com/getsentry/sentry-javascript/issues/2978). So I wonder if we could just replace `const global = getGlobalObject(); global.blahblahblah()` with `globalThis.blahblahblah()` and skirt the issue entirely. WDYT?

lforst replied:
Removed the changes to global from this PR. Tests seem to be working again. Gremlins.

lobsterkatie replied:
LOL. Great. I think the `global`/`globalThis` convo might be worth having, but I agree - separate PR for that.

lforst commented on packages/gatsby/test/globals.d.ts:
```diff
@@ -0,0 +1,10 @@
+ * replaced.
+ */
+
+declare const __DEBUG_BUILD__: boolean;
```
The million extra files annoy me to the max. Let me investigate a bit more...

lobsterkatie commented on packages/utils/jest.config.js:
```diff
@@ -6,4 +6,10 @@ module.exports = {
     '^.+\\.ts$': 'ts-jest',
     '^.+\\.js$': 'ts-jest',
   },
+  globals: {
+    'ts-jest': {
+      tsconfig: '<rootDir>/tsconfig.test.json',
+    },
+    __DEBUG_BUILD__: true,
+  },
```
How come this needs to be repeated here, given that it already exists in the main jest config?

lforst replied:
Yup that's unnecessary. I think I used the utils tests to debug what I need to change. Removed. d68d3e74444e4be4aabdc8d3223f87211e304710

lobsterkatie commented on rollup/plugins/bundlePlugins.js:
```diff
@@ -46,9 +46,12 @@ export function makeLicensePlugin(title) {
 export function makeIsDebugBuildPlugin(includeDebugging) {
   return replace({
-    // __SENTRY_DEBUG__ should be save to replace in any case, so no checks for assignments necessary
+    // __DEBUG_BUILD__ and __SENTRY_DEBUG__ are save to replace in any case, so no checks for assignments necessary
```
```suggestion
    // __DEBUG_BUILD__ and __SENTRY_DEBUG__ are safe to replace in any case, so no checks for assignments necessary
```

lobsterkatie commented on rollup/npmHelpers.js:
```diff
@@ -84,6 +86,7 @@ export function makeBaseNPMConfig(options = {}) {
       removeESLintCommentsPlugin,
       removeBlankLinesPlugin,
       extractPolyfillsPlugin,
+      debugBuildStatementReplacePlugin,
```
In this case I don't think it makes a behavioral difference, but can this go between the sucrase and constToVar plugins? (It's more reflective of the order in which things actually happen.)

lforst replied:
Done. 935f9b19edc4fdf2865c1c4b76b273bcc4148f2c

lobsterkatie commented on packages/types/src/index.ts:
```diff
@@ -1,3 +1,5 @@
+import './globals';
```
This may be the fishiness you were talking about, but... how does this work?

lforst replied:
Take this with a grain of salt because I don't know a 100% either.

Fwiw we could also put `declare global { const __DEBUG_BUILD__: boolean; }` right here in the index.ts file. That would also work. However, (I think) we want it in a separate file so we need to somehow import it so that the ambient declaration `global` gets picked up when packages import from the types package, because the `"types"` entrypoint only points to `index.d.ts`.

`import './globals';` worked for consuming packages so I just used that. `import [type] {} from './globals'` did not work. Out of desperation I just tried `export type {} from './globals';`. That also seems to work and it's a little bit more in line with the other stuff we have in the package so I'm changing it to that (b2928d991f16002de5855d211691d18e5bcc0f0b). 

So yeah... that's how *I think* this whole thing works

lobsterkatie replied:
> `export type {} from './globals';`

So, in other words, this is a dummy export just to cause TS to load the `globals` file, which shoves `__DEBUG_BUILD__` into the ambient global namespace? (Which would explain how the import worked, also.)

lobsterkatie commented on packages/integrations/scripts/buildBundles.sh:
```diff
@@ -6,7 +6,7 @@ for filepath in ./src/*; do
     # integration) and the flags file is just a helper for including or not including debug logging, whose contents gets
     # incorporated into each of the individual integration bundles, so we can skip them both here.
-    if [[ $file == "index.ts" || $file == "flags.ts" ]]; then
+    if [[ $file == "index.ts" || $file == "globals.d.ts" ]]; then
```
Given that you took out all of the package-level `globals` files, do you still need this?

lforst replied:
Yup you're right. Removed. f84a2157d6389b109b28b2d7bf0ee285a6bcc4d1

lobsterkatie commented on .gitignore:
```diff
@@ -10,7 +10,6 @@ build/
 dist/
 coverage/
 scratch/
-*.d.ts
```
Rather than do this, I'd rather just carve out an exception in the types package (as we do in the utils package). In general, these are build artifacts, and shouldn't be committed to version control.

lforst replied:
I added this back because this change became unnecessary with the latest changes f84a2157d6389b109b28b2d7bf0ee285a6bcc4d1

lobsterkatie commented:
The ultimate reason we're doing this is so that the code actually implementing the logger gets treeshaken alongside all of the log statements, right? If so, can you please add a note to that effect in the PR description?

lforst commented:
> The ultimate reason we're doing this is so that the code actually implementing the logger gets treeshaken alongside all of the log statements, right? If so, can you please add a note to that effect in the PR description?

Added to description. I think at some point in between the "function to const PR" and v7 we switched from the Logger Class to just a logger object + functions so I don't know what this PR actually does with logger code compared to then. I still firmly believe that from now on, logger statements are much easier to tree shake so this is what I wanted the PR description to be about.

lobsterkatie commented:
> > The ultimate reason we're doing this is so that the code actually implementing the logger gets treeshaken alongside all of the log statements, right? If so, can you please add a note to that effect in the PR description?
> 
> Added to description. I think at some point in between the "function to const PR" and v7 we switched from the Logger Class to just a logger object + functions so I don't know what this PR actually does with logger code compared to then. I still firmly believe that from now on, logger statements are much easier to tree shake so this is what I wanted the PR description to be about.

Ah, got it. 

> We hope this fixes tree shaking the logger for downstream consumers.

If we don't know, maybe best not to bring it up, actually. (That or we should actually test it out and see.) WDYT?

As an aside, on the topic of treeshaking the logger itself: In the process of converting from a class to functions, [this `else` somehow got introduced](https://github.com/getsentry/sentry-javascript/blob/72ce5dab0bead8e8797d3a27c0b0e70e8bdf6f48/packages/utils/src/logger.ts#L94-L95), but it shouldn't be there. It doesn't explain why the logger stopped getting treeshaken (as it happened after your function-to-constant change) but it would, I believe, prevent the treeshaking from ever working (with or without the changes in this PR). Probably should be fixed in a separate PR, though, and that's probably where we should be talking about the question of treeshaking the logger. So, yeah, actually, I think scratch my original request to add it to the PR description here. Let's hold off, talk about it there, and have tested it so we can say for sure whether or not it works.

lobsterkatie commented on rollup/bundleHelpers.js:
```diff
@@ -157,7 +157,7 @@ export function makeBundleConfigVariants(baseConfig, options = {}) {
         entryFileNames: chunkInfo => `${baseConfig.output.entryFileNames(chunkInfo)}.debug.min.js`,
       },
-      plugins: [terserPlugin],
+      plugins: [includeDebuggingPlugin, terserPlugin],
```
This isn't super in scope for this PR (and it was my mistake), so I've fixed it in https://github.com/getsentry/sentry-javascript/pull/5192.

lobsterkatie commented on packages/types/src/globals.ts:
```diff
@@ -0,0 +1,13 @@
+ * This file defines flags and constants that can be modified during compile time in order to facilitate tree shaking
+ * for users.
+ *
+ * "Magic strings" like `__SENTRY_DEBUG__` must be replaced with actual values or safe statements during our build process.
```
Should this be `__DEBUG_BUILD__`?

lforst replied:
Yup thanks! 3d63759a0cfbebaadd98784fb0bca35e8811d60e

lobsterkatie commented on CONTRIBUTING.md:
```diff
@@ -61,16 +61,14 @@ Pro tip: If any of your breakpoints are in code run by multiple tests, and you r
 These flags serve two purposes:
 
 1. They enable us to remove debug code for our production browser bundles.
 2. Enable users to tree-shake Sentry debug code for their production builds.
 
-These debug flags need to be declared in each package individually and must not be imported across package boundaries, because some build tools have trouble tree-shaking imported guards.
-As a convention, we define debug flags in a `flags.ts` file in the root of a package's `src` folder.
-The `flags.ts` file will contain "magic strings" like `__SENTRY_DEBUG__` that may get replaced with actual values during our, or the user's build process.
-Take care when introducing new flags - they must not throw if they are not replaced.
+Flags will be replaced with booleans for our browser bundles. In cjs and esm builds, they will be replaced by statements that do not throw.
```
It won't let me actually make a suggestion here, but if I could, I'd suggest:

```md
Throughout the codebase, you will find the `__DEBUG_BUILD__` flag guarding various code sections. This flag serves two purposes:

1. It enables us to remove debug code from our minified CDN bundles during build, by replacing the flag with `false` before tree-shaking occurs.
2. It enables users to remove Sentry debug code from their production bundles during their own build. When we build our npm packages, we replace the flag with `(typeof __SENTRY_DEBUG__ === 'undefined' || __SENTRY_DEBUG__)`. If the user does nothing, this evaluates to `true` and logging is included. But if the user runs their own replacement during build (again replacing the flag with `false`), the build will tree-shake the logging away, just as our bundle builds do.

Note that the replacement flag, `__SENTRY_DEBUG__`, is different from the original flag . This is necessary because the replacement plugin runs twice, at two different stages of the build, and we don't want to run a replacement on our replacement (as would happen if we reused `__DEBUG_BUILD__`).
```

lforst replied:
Great change. Thank you! a4b07d329c142b5462c276de7f4da2a9449e961a

lobsterkatie reviewed (commented):
Just a few more small things, and then I think this will be ready to go!

lforst commented on packages/gatsby/.gitignore:
```diff
@@ -0,0 +1,2 @@
+gatsby-browser.d.ts
+gatsby-node.d.ts
```
Correct, thanks! 690c68ce98f514c5be6444f1a44833bebb614ef0

lforst commented on packages/types/src/index.ts:
```diff
@@ -29,6 +29,7 @@ export type { Event, EventHint } from './event';
 export type { EventProcessor } from './eventprocessor';
 export type { Exception } from './exception';
 export type { Extra, Extras } from './extra';
+import './globals';
```
This PR is a mess I apologize. Fixed in 613e2f409cfe3eb2b2f39de5f5d56b1541630ee6

lobsterkatie replied:
I seem to recall you catching me in at least two fairly dumb mistakes* some time within your first week or so of working at Sentry, so...

\*To be clear, I'm not saying your mistakes are dumb, but mine def were, and I was grateful you caught them before I merged them in!

lforst commented:
@lobsterkatie I realize this PR has been a pain. Thanks for sticking with me!

lobsterkatie commented on rollup/plugins/bundlePlugins.js:
```diff
@@ -46,9 +46,12 @@ export function makeLicensePlugin(title) {
     values: {
+      // Flags in current package
+      __DEBUG_BUILD__: includeDebugging,
+      // Flags in dependencies
```
```suggestion
      // Flags in built monorepo dependencies, from which the bundle pulls
```

lobsterkatie commented on rollup/plugins/npmPlugins.js:
```diff
@@ -127,4 +127,20 @@ export function makeRemoveBlankLinesPlugin() {
+/**
+ * Creates a plugin to replace all instances of "__DEBUG_BUILD__" with a safe statement that
+ * a) evaluates to `true`
+ * b) can easily be modified by bundlers to evaluate to false, facilitating the treeshaking of logger code.
```
```suggestion
 * b) can easily be modified by our users' bundlers to evaluate to false, facilitating the treeshaking of logger code.
```
